src/main/kotlin/io/bazel/kotlin/builder/tasks/BazelWorker.kt

```diff
@@ -56,107 +63,164 @@ interface CommandLineProgram {
-          if (!isWorker && mnemonic.isNotEmpty()) {
-            output.printf(
-                "HINT: %s will compile faster if you run: " + "echo \"build --strategy=%s=worker\" >>~/.bazelrc\n",
-                mnemonic, mnemonic
-            )
-          }
-          try {
-            return Files.readAllLines(flagFile, UTF_8)
-          } catch (e: IOException) {
-            throw RuntimeException(e)
-          }
-        }
+/** Defines the common worker interface. */
+interface Worker {
+  fun run(args: List<String>): Int
+}
+
+class WorkerIO(
+  val input: InputStream,
+  val output: PrintStream,
+  val execution: ByteArrayOutputStream,
+  private val restore: () -> Unit
+) : Closeable {
+  companion object {
+    fun open(): WorkerIO {
+      val stdErr = System.err
+      val stdIn = BufferedInputStream(System.`in`)
+      val stdOut = System.out
+      val inputBuffer = ByteArrayInputStream(ByteArray(0))
+      val execution = ByteArrayOutputStream()
+      val outputBuffer = PrintStream(execution)
+
+      // delegate the system defaults to capture execution information
+      System.setErr(outputBuffer)
+      System.setOut(outputBuffer)
+      System.setIn(inputBuffer)
+
+      return WorkerIO(stdIn, stdOut, execution) {
+        System.setOut(stdOut)
+        System.setIn(stdIn)
+        System.setErr(stdErr)
       }
     }
-    return args
+
+    fun noop(): WorkerIO {
+      val inputBuffer = ByteArrayInputStream(ByteArray(0))
+      val execution = ByteArrayOutputStream()
+      val outputBuffer = PrintStream(execution)
+      return WorkerIO(inputBuffer, outputBuffer, execution) {}
+    }
+  }
+
+  override fun close() {
+    restore.invoke()
   }
 }
 
-class InvocationWorker(
-  private val delegate: CommandLineProgram,
-  private val buffer: ByteArrayOutputStream
-) {
+/** PersistentWorker follows the Bazel worker protocol and executes a CommandLineProgram. */
+class PersistentWorker(
+  private val io: WorkerIO,
+  private val program: CommandLineProgram
+) : Worker {
+  val logger = Logger.getLogger(PersistentWorker::class.java.canonicalName)
+
+  enum class Status {
+    OK, INTERRUPTED, ERROR
+  }
+
+  override fun run(args: List<String>): Int {
+    while (true) {
+      val request = WorkRequest.parseDelimitedFrom(io.input) ?: continue
 
-  fun invoke(args: List<String>): Pair<Int, String> = WorkingDirectoryContext.newContext()
-      .use { wdCtx ->
-        return try {
-          delegate.apply(wdCtx.dir, args)
-        } catch (e: RuntimeException) {
-          if (e.wasInterrupted()) BazelWorker.INTERRUPTED_STATUS
-          else BazelWorker.ERROR_STATUS.also {
-            System.err.println(
-                "ERROR: Worker threw uncaught exception with args: ${args}"
-            )
-            e.printStackTrace(System.err)
+      val (status, exit) = WorkingDirectoryContext.newContext()
+        .runCatching {
+          request.argumentsList
+            ?.let {
+              maybeExpand(it)
```

Comment:
nit: I'd make these one-liners. But this isn't wrong.

Reply:
I get curly bracket blindness when the density is too tight.